### DISCUSSION TITLE
Is is it possible to create a passkey on the same device you are using?

### DISCUSSION THREAD

**Discussion by treeder** [MasterKale/SimpleWebAuthn#539] [General]
I've got it working on desktop and it will ask my phone to create the passkey. But if I'm on my phone, it gives me the options below, so I'm unable to create one. 

![Screenshot_20240321-123640](https://github.com/MasterKale/SimpleWebAuthn/assets/75826/777350ea-1285-4a89-a11c-2146f05cf9d3)

**Comment by treeder** [comment]
I removed the whole `authenticatorSelection` and it now it seems to work properly, allowing me to create an passkey on the same device.

**Comment by MasterKale** [comment]
That looks like what I'd expect Android to pop up if `residentKey: "discouraged"` is being specified. The "new" WebAuthn functionality that allows registering synced passkeys requires at least `residentKey: "preferred"` if not `residentKey: "required"`.

Note that preferring or requiring resident keys here will have a potentially unintended consequence of **security keys** creating and storing a "discoverable credential" (new name, same concept as resident keys) internally. As security keys have limited internal storage this can be seen as a bad thing by security key users. If you imagine you'll have a lot of users using security keys then you may want to consider alternative registration flows, one for "typical registration of just about anything" that sets `residentKey: "required"` to get the most consistent synced passkeys experience; and a second one specifically for security keys that specifies `residentKey: "discouraged"` with `authenticatorAttachment: "cross-platform"`. Just my $0.02 from being in these trenches for so long ✌️

**Reply by treeder** [reply]
I definitely wanted cross-platform and tried various combinations of residentKey and userVerification but couldn't get it to work like I wanted on both desktop and my phone. 

Removing that whole authenticatorSelection works just how I want it though giving me all the options.

**Reply by treeder** [reply]
I just realized, I wrote authenticatorAttachment in my last response, I meant the authenticatorSelection block (updated).

**Reply by treeder** [reply]
Ehh, ok, I think I'm wrong, I believe it's making "singleDevice" keys now...

**Reply by treeder** [reply]
Ok, so here's what I get for various options (on desktop):

If I remove `authenticatorAttachment` and make `residentKey` required, I **think** it's doing what it should. Is that what you are suggesting?

If I specify platform or cross-platform, it doesn't do what I want.

**Reply by MasterKale** [reply]
> If I remove `authenticatorAttachment` and make `residentKey` required, I **think** it's doing what it should. Is that what you are suggesting?

Wait the issue was that you were seeing the prompt in the screenshot above while your options contained `authenticatorAttachment: "cross-platform"`? Then I'd say that's things working as intended. **If you want to register the user's local device then that's `authenticatorAttachment: "platform"`.**

The behavior you're seeing when you omit `authenticatorAttachment` is the phone wanting you to register the local platform authenticator because it's not being told by the RP (you) what kind of authenticator to get the user to register.

Does this make sense?

**Reply by treeder** [reply]
Ok, so here's what happens with various options:

## Option 1: no `authenticatorAttachment`

```js
  authenticatorSelection: {
                    residentKey: 'required',
                    userVerification: 'preferred',
                    // authenticatorAttachment: 'platform', // 'cross-platform',
                },
```


desktop:

![Screenshot 2024-03-25 8 06 29 AM](https://github.com/MasterKale/SimpleWebAuthn/assets/75826/9de40a82-4542-4e7e-aff3-ebcc736c79dd)

mobile:
![Screenshot_20240325-080647-EDIT](https://github.com/MasterKale/SimpleWebAuthn/assets/75826/1b892c69-ced0-433a-b729-6a36f9a7d3d8)

And if I click "use a different device":

![Screenshot_20240325-080653](https://github.com/MasterKale/SimpleWebAuthn/assets/75826/c50af256-a828-4d02-9e44-425521cf1b59)

So that option seems good and it appears to make a `multiDevice` synced key which is good. 


## Option 2: "platform"

```js
 authenticatorSelection: {
                    residentKey: 'required',
                    userVerification: 'preferred',
                    authenticatorAttachment: 'platform', // 'cross-platform',
                },
```

desktop:

![Screenshot 2024-03-25 8 07 35 AM](https://github.com/MasterKale/SimpleWebAuthn/assets/75826/f35777a7-2448-4a96-9a22-1d38a65d32ac)

mobile:

![Screenshot_20240325-080937](https://github.com/MasterKale/SimpleWebAuthn/assets/75826/30c97342-f1a7-4e64-8faa-bde3c596c0a1)

So this one won't work on desktop and mobile only gives one option (ie: the local device), there is "use a different device" option. 

## Option 3: cross-platform

```js
 authenticatorSelection: {
                    residentKey: 'required',
                    userVerification: 'preferred',
                    authenticatorAttachment: 'cross-platform',
                },
```

desktop:

![Screenshot 2024-03-25 8 17 42 AM](https://github.com/MasterKale/SimpleWebAuthn/assets/75826/ce502880-57a1-4376-88b8-f52380faa886)

Notice no local device option. 

mobile:

![Screenshot_20240325-081836](https://github.com/MasterKale/SimpleWebAuthn/assets/75826/4ccc8d8a-524b-49c7-ac64-292365bcc480)

Again, no local device option.

**Reply by treeder** [reply]
Based on that, I think the best option is to not define the `authenticatorAttachment`.

**Reply by MasterKale** [reply]
> desktop:
> 
> ![Screenshot 2024-03-25 8 07 35 AM](https://private-user-images.githubusercontent.com/75826/316552895-f35777a7-2448-4a96-9a22-1d38a65d32ac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEzODI4ODYsIm5iZiI6MTcxMTM4MjU4NiwicGF0aCI6Ii83NTgyNi8zMTY1NTI4OTUtZjM1Nzc3YTctMjQ0OC00YTk2LTlhMjItMWQzOGE2NWQzMmFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI1VDE2MDMwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM1YWMxMWFhOGY3Mzg3YTRkNmI3ODRiNmQwNmE0NGNhOWU2OGYzNjRjMDUyNzkxOGViOTFhODY3Mzg1YjI4MjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9o-q-u5wYD-tE8se-zR2t8p3ywXfpSUNXdvPDQf-2kg)

This looks like you don't have a local platform authenticator installed. Are you on Linux by chance? Just curious.

But yeah, if you want to support the greatest number of use cases then it's best to omit `authenticatorAttachment`, especially now that hybrid registration is an option (so users in your similar situation can use their phone to register something backed by a platform authenticator and thus probably synced.)
